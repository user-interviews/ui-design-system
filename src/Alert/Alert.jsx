import React, { useEffect } from 'react';
import PropTypes from 'prop-types';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faBullhorn,
  faCircle,
  faCheck,
  faExclamationTriangle,
  faInfo,
  faTimes,
} from '@fortawesome/pro-solid-svg-icons';

import './Alert.scss';
import classNames from 'classnames';

export const MessageTypes = {
  SUCCESS: 'success',
  INFO: 'info',
  ANNOUNCEMENT: 'announcement',
  WARNING: 'warning',
  ERROR: 'error',
};

const getAlertIcon = (type) => {
  switch (type) {
    case MessageTypes.SUCCESS:
      return (
        <span className="fa-layers fa-fw">
          <FontAwesomeIcon icon={faCircle} transform="grow-8" />
          <FontAwesomeIcon icon={faCheck} transform="shrink-4" />
        </span>
);
    case MessageTypes.INFO:
      return (
        <span className="fa-layers fa-fw">
          <FontAwesomeIcon icon={faCircle} transform="grow-8" />
          <FontAwesomeIcon icon={faInfo} transform="shrink-4" />
        </span>
      );
    case MessageTypes.ANNOUNCEMENT:
      return (<FontAwesomeIcon icon={faBullhorn} transform="grow-2" />);
    case MessageTypes.WARNING:
      return (<FontAwesomeIcon icon={faExclamationTriangle} transform="grow-2" />);
    case MessageTypes.ERROR:
      return (<FontAwesomeIcon icon={faExclamationTriangle} transform="grow-2" />);
    default:
      return null;
  }
};

const AUTO_DISMISS_TIMEOUT = 5000;

const getAlertClassName = (type) => {
  if (!Object.values(MessageTypes).includes(type)) {
    throw new TypeError(`Unexpected type ${type} used for an alert.`);
  }

  return `Alert Alert-${type}`;
};

function Alert(props) {
  const { autoDismiss, id, onDismiss } = props;

  useEffect(() => {
    let timeout;
    if (autoDismiss) {
      timeout = setTimeout(() => (onDismiss(id)), AUTO_DISMISS_TIMEOUT);
    }
    return () => {
      clearTimeout(timeout);
    };
  }, [autoDismiss, onDismiss, id]);

  return (
    <div
      className={getAlertClassName(props.type)}
      style={props.removeBorderLeft ? { borderLeft: 'none' } : null}
    >
      <div className="Alert__icon">
        {getAlertIcon(props.type)}
      </div>
      <div className="Alert__content">
        {
          props.title && (
            <div className="Alert__title">
              {props.title}
            </div>
          )
        }
        <div
          className="Alert__message"
          // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
          tabIndex={props.type === MessageTypes.ERROR ? -1 : undefined}
        >
          {props.message}
        </div>
      </div>
      {
        props.action && (
        <div className="Alert__action">
          { React.isValidElement(props.action) ?
            props.action : (
              <a
                className={classNames(`Alert-${(props.type)}`, 'primary-action')}
                href={props.action.url}
                rel="noopener noreferrer"
              >
                {props.action.content}
              </a>
          )}
        </div>
      )
}
      {
        props.onDismiss && (
          <div className="Alert__close">
            <button
              aria-label={`close ${props.type}`}
              className="close"
              type="button"
              onClick={() => props.onDismiss(props.id)}
            >
              <FontAwesomeIcon icon={faTimes} />
            </button>
          </div>
        )
      }
    </div>
  );
}

Alert.propTypes = {
  /**
   Creates a CTA button on the Alert
  */
  action: PropTypes.oneOfType([PropTypes.object, PropTypes.node]),
  /**
   Determines whether the Alert will disappear automatically
  */
  autoDismiss: PropTypes.bool,
  id: PropTypes.string.isRequired,
  message: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
  removeBorderLeft: PropTypes.bool,
  title: PropTypes.string,
  /**
   One of the MessageTypes
  */
  type: PropTypes.oneOf(Object.values(MessageTypes)).isRequired,
  onDismiss: PropTypes.func,
};

Alert.defaultProps = {
  action: undefined,
  autoDismiss: false,
  removeBorderLeft: false,
  title: undefined,
  onDismiss: undefined,
};

export default React.memo(Alert);
