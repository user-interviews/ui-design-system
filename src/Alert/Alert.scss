@import '../../scss/theme';

@mixin primary-action($bg-color: none, $bg-color-hover: $ux-gray-300, $color: var(--synth-unselected-neutral)) {
  @include synth-font-type-30--bold;
  background-color: $bg-color;
  border: none;
  border-left: unset;
  border-radius: var(--ux-border-radius);
  color: $color;
  margin: $synth-spacing-2 $synth-spacing-4 $synth-spacing-2 0;
  padding: $synth-spacing-2 $synth-spacing-4;
  text-decoration: none;
  white-space: nowrap;

  &:hover {
    background-color: $bg-color-hover;
  }
}

@mixin close-action($color: $ux-gray-700, $color-hover: $ux-gray-800) {
  background-color: unset;
  border: none;
  color: $color;

  &:hover {
    color: $color-hover
  }
}

.Synthesized {
  .Alert {
    display: grid;
    grid-template-columns: 32px auto auto 32px;
    grid-template-rows: auto auto;
    grid-template-areas:
      'icon content content close'
      'icon action action close'
      ;
  
    @media (min-width: 450px) {
      grid-template-areas:
      'icon content action close'
      'icon content action close'
      ;
    }
  
    background-color: var(--synth-alert-bg-neutral);
    border-radius: var(--ux-border-radius);
    color: var(--synth-text-neutral);
    margin-bottom: $synth-spacing-4;
    padding: $synth-spacing-4;
    position: relative;
  
    &__icon {
      grid-area: icon;
    }
  
    &__content {
      grid-area: content;
      margin-right: $synth-spacing-4;
    }
  
    &__action {
      grid-area: action;
      display: flex;
      align-items: center;
      justify-content: flex-start;
  
      @media (min-width: 450px) {
        justify-content: flex-end;
      }
    }
  
    &__close {
      grid-area: close;
      display: flex;
      align-items: flex-start;
      justify-content: flex-end;
    }
  
    &__title {
      @include synth-font-type-30--bold;
      margin-bottom: 4px;
    }
  
    &__message {
      @include synth-font-type-30;
      margin-bottom: 0;
    }
  }
  
  .Alert-success {
    border-left: 8px solid var(--synth-accent-green);
  
    .close {
      @include close-action;
    }
  
    .primary-action {
      @include primary-action;
    }
  
    .Alert__icon {
      color: var(--synth-accent-green);
      
      .fa-check {
        color: $ux-gray-100;
      }
    }
  }
  
  .Alert-info {
    border-left: 8px solid var(--synth-hyperlink-color);
  
    .close {
      @include close-action;
    }
  
    .primary-action {
      @include primary-action;
    }
  
    .Alert__icon {
      color: $ux-blue-300;
      
      .fa-info {
        color: $ux-blue-600;
      }
    }
  }
  
  .Alert-announcement {
    border-left: 8px solid var(--synth-hyperlink-color);
  
    .close {
      @include close-action;
    }
  
    .primary-action {
      @include primary-action;
    }
  }
  
  .Alert-feature {
    border: 2px solid $ux-teal-600;
  
    .close {
      @include close-action;
    }
  
    .primary-action {
      @include primary-action;
    }
  
    .Alert__icon {
      color: $ux-teal-600;
    }
  }
  
  .Alert-warning {
    border-left: 8px solid var(--synth-warning-amber-alternate);
  
    .close {
      @include close-action;
    }
  
    .primary-action {
      @include primary-action;
    }
  
    .Alert__icon {
      color: var(--synth-warning-amber-alternate);
    }
  }
  
  .Alert-error {
    border-left: 8px solid $ux-red-400;
  
    .close {
      @include close-action;
    }
  
    .primary-action {
      @include primary-action;
    }
  
    .Alert__icon {
      color: $ux-red;
    }
  }
}

.Alert {
  display: grid;
  grid-template-columns: 32px auto auto 32px;
  grid-template-rows: auto auto;
  grid-template-areas:
    'icon content content close'
    'icon action action close'
    ;

  @media (min-width: 450px) {
    grid-template-areas:
    'icon content action close'
    'icon content action close'
    ;
  }

  font-weight: 500;
  border-radius: var(--ux-border-radius);
  margin-bottom: $synth-spacing-4;
  padding: $synth-spacing-4;
  position: relative;

  &__icon {
    grid-area: icon;
  }

  &__content {
    grid-area: content;
    margin-right: $synth-spacing-4;
  }

  &__action {
    grid-area: action;
    display: flex;
    align-items: center;
    justify-content: flex-start;

    @media (min-width: 450px) {
      justify-content: flex-end;
    }
  }

  &__close {
    grid-area: close;
    display: flex;
    align-items: flex-start;
    justify-content: flex-end;
  }

  &__title {
    @include synth-font-type-30--bold;
    margin-bottom: 4px;
  }

  &__message {
    @include synth-font-type-30;
    margin-bottom: 0;
  }
}

.Alert-success {
  background-color: $ux-green-100;
  color: $ux-green-800;
  border-left: 8px solid $ux-green-400;

  .close {
    @include close-action;
  }

  .primary-action {
    @include primary-action;
  }

  .Alert__icon {
    color: $ux-green-400;
    
    .fa-check {
      color: $ux-green-800;
    }
  }
}

.Alert-info {
  background-color: $ux-blue-100;
  color: $ux-blue-800;
  border-left: 8px solid $ux-blue-300;

  .close {
    @include close-action($ux-blue-600, $ux-blue-800)
  }

  .primary-action {
    @include primary-action;
  }

  .Alert__icon {
    color: $ux-blue-300;
    
    .fa-info {
      color: $ux-blue-600;
    }
  }
}

.Alert-announcement {
  background-color: $ux-blue-100;
  color: $ux-blue-700;
  border-left: 8px solid $ux-blue-300;

  .close {
    @include close-action;
  }

  .primary-action {
    @include primary-action;
  }
}

.Alert-feature {
  background-color: $ux-teal-100;
  color: $ux-teal-800;
  border: 2px solid $ux-teal-600;

  .close {
    @include close-action;
  }

  .primary-action {
    @include primary-action;
  }
}

.Alert-feature {
  background-color: $ux-teal-100;
  color: $ux-teal-800;
  border: 2px solid $ux-teal-600;

  .close {
    @include close-action($ux-teal-600, $ux-teal-800)
  }

  .primary-action {
    @include primary-action($ux-teal-600, $ux-teal-700, $ux-white)
  }
}

.Alert-warning {
  background-color: $ux-yellow-100;
  color: $ux-yellow-900;
  border-left: 8px solid $ux-yellow-600;

  .close {
    @include close-action;
  }

  .primary-action {
    @include primary-action;
  }

  .Alert__icon {
    color: $ux-yellow-700;
  }
}

.Alert-error {
  background-color: $ux-red-100;
  color: $ux-red-800;
  border-left: 8px solid $ux-red-400;

  .close {
    @include close-action;
  }

  .primary-action {
    @include primary-action;
  }

  .Alert__icon {
    color: $ux-red;
  }
}
