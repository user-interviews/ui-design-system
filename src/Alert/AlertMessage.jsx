import React, { useEffect } from 'react';
import PropTypes from 'prop-types';

import './AlertMessage.scss';

export const MessageTypes = {
  SUCCESS: 'success',
  INFO: 'info',
  WARNING: 'warning',
  ERROR: 'error',
};

const AUTO_DISMISS_TIMEOUT = 5000;

const alertMessageClassName = (type) => {
  if (!Object.values(MessageTypes).includes(type)) {
    throw new TypeError(`Unexpected type ${type} used for a alert message.`);
  }

  return `AlertMessage AlertMessage-${type}`;
};

function AlertMessage(props) {
  const { autoDismiss, id, onDismiss } = props;

  useEffect(() => {
    let timeout;
    if (autoDismiss) {
      timeout = setTimeout(() => (onDismiss(id)), AUTO_DISMISS_TIMEOUT);
    }
    return () => {
      clearTimeout(timeout);
    };
  }, [autoDismiss, onDismiss, id]);

  return (
    <div className={alertMessageClassName(props.type)}>
      {
        props.onDismiss && (
          <button
            className="close"
            type="button"
            onClick={() => props.onDismiss(props.id)}
          >
            &times;
          </button>
        )
      }
      {
        props.title && (
          <p className="AlertMessage__title">{props.title}</p>
        )
      }
      {props.message}
    </div>
  );
}

AlertMessage.propTypes = {
  autoDismiss: PropTypes.bool,
  id: PropTypes.string.isRequired,
  message: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
  title: PropTypes.string,
  type: PropTypes.string.isRequired,
  onDismiss: PropTypes.func,
};

AlertMessage.defaultProps = {
  autoDismiss: false,
  title: undefined,
  onDismiss: undefined,
};

export default React.memo(AlertMessage);
