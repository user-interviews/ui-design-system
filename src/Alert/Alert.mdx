import { ArgsTable, Story, Preview } from '@storybook/addon-docs/blocks';
import Alert from './Alert'

# Alerts

Alerts give users updates or changes to system status. Providing immediate feedback is important for
building trust.  While alerts are a great way of communicating with users, they are disruptive and
should be used cautiously.

<Preview>
  <Story id="design-system-alert--success"/>
</Preview>

<ArgsTable of={Alert} />

## Anatomy

- **Icon:** Allow users to identify the type of alert at a glance
- **Title:** Summary of alert topic 
- **Message:** Provide details or actionable step for user
- **CTA** ***(currently WIP):*** To provide more details or an escape so users can finish the task they were doing.
- **Close icon:** Closes the alert

## Stories

### Success

- Used when something went as planned. (e.g. When you've saved your progress in the builder.)

<Preview>
  <Story id="design-system-alert--success" />
</Preview>

### Info

- Information that needs to be referenced often or is very important. These alerts are within the
page and are not affected by an event, whereas success/error alerts are at the top of the page
based on an event. These alerts will always be blue!

<Preview>
  <Story id="design-system-alert--info" />
</Preview>

### Error

- Used when something did not go as planned, and action is needed to proceed. (e.g. If you leave
the incentive field blank in the builder and try to launch, you receive an error. You must
fill in the field to launch a project.)

<Preview>
  <Story id="design-system-alert--error" />
</Preview>

### Warning

- Messages used for general information on a more temporary level, not based on the user
performing an action. (e.g For COVID, we show a yellow alert to encourage researcher to
consult CDC guidelines for in person sessions.

<Preview>
  <Story id="design-system-alert--warning" />
</Preview>


### With Dismiss

- Used in cases where a user would want to dismiss or remove an alert from view.

<Preview>
  <Story id="design-system-alert--with-dismiss" />
</Preview>
