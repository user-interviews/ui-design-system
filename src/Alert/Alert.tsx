import React, { useEffect, type ReactNode, type ReactElement } from 'react';

import type { IconDefinition } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faBullhorn,
  faCircle,
  faCheck,
  faExclamationTriangle,
  faInfo,
  faTimes,
} from '@fortawesome/pro-solid-svg-icons';

import './Alert.scss';
import classNames from 'classnames';

export enum MessageTypes {
  SUCCESS = 'success',
  INFO = 'info',
  ANNOUNCEMENT = 'announcement',
  FEATURE = 'feature',
  WARNING = 'warning',
  ERROR = 'error',
}

const getAlertIcon = (type: MessageTypes) => {
  switch (type) {
    case MessageTypes.SUCCESS:
      return (
        <span className="fa-layers fa-fw">
          <FontAwesomeIcon icon={faCircle as IconDefinition} transform="grow-8" />
          <FontAwesomeIcon icon={faCheck as IconDefinition} transform="shrink-4" />
        </span>
);
    case MessageTypes.INFO:
      return (
        <span className="fa-layers fa-fw">
          <FontAwesomeIcon icon={faCircle as IconDefinition} transform="grow-8" />
          <FontAwesomeIcon icon={faInfo as IconDefinition} transform="shrink-4" />
        </span>
      );
    case MessageTypes.ANNOUNCEMENT:
      return (<FontAwesomeIcon icon={faBullhorn as IconDefinition} transform="grow-2" />);
    case MessageTypes.FEATURE:
      return (<FontAwesomeIcon icon={faBullhorn as IconDefinition} transform="grow-2" />);
    case MessageTypes.WARNING:
      return (<FontAwesomeIcon icon={faExclamationTriangle as IconDefinition} transform="grow-2" />);
    case MessageTypes.ERROR:
      return (<FontAwesomeIcon icon={faExclamationTriangle as IconDefinition} transform="grow-2" />);
    default:
      return null;
  }
};

const AUTO_DISMISS_TIMEOUT_SUCCESS = 3000;
const AUTO_DISMISS_TIMEOUT_DEFAULT = 5000;

const getAlertClassName = (type: MessageTypes) => {
  if (!Object.values(MessageTypes).includes(type)) {
    throw new TypeError(`Unexpected type ${type} used for an alert.`);
  }

  return `Alert Alert-${type}`;
};

type AlertProps = {
/**
   Creates a CTA button on the Alert
  */
  action?: ReactElement | {
    url: string;
    content: ReactNode;
  };
  /**
    Specifies where to open the linked document
  */
  actionTarget?: string;
  /**
   Determines whether the Alert will disappear automatically
  */
  autoDismiss?: boolean;
  id: string;
  message: string | ReactNode;
  removeBorderLeft?: boolean;
  title?: string;
  /**
   One of the MessageTypes
  */
  type: MessageTypes;
  onDismiss?: (arg0: string) => void;
}

function Alert({
  action,
  actionTarget,
  autoDismiss,
  id,
  message,
  onDismiss,
  removeBorderLeft,
  title,
  type,
}: AlertProps) {
  useEffect(() => {
    let timeout;
    if (autoDismiss && onDismiss) {
      timeout = setTimeout(() => (onDismiss(id)),
      type === MessageTypes.SUCCESS ?
      AUTO_DISMISS_TIMEOUT_SUCCESS : AUTO_DISMISS_TIMEOUT_DEFAULT);
    }
    return () => {
      clearTimeout(timeout);
    };
  }, [autoDismiss, onDismiss, id, type]);

  return (
    <div
      className={getAlertClassName(type)}
      style={removeBorderLeft ? { borderLeft: 'none' } : undefined}
    >
      <div className="Alert__icon">
        {getAlertIcon(type)}
      </div>
      <div className="Alert__content">
        {
          title && (
            <div className="Alert__title">
              {title}
            </div>
          )
        }
        <div
          className="Alert__message"
          // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
          tabIndex={type === MessageTypes.ERROR ? -1 : undefined}
        >
          {message}
        </div>
      </div>
      {
        action && (
        <div className="Alert__action">
          { /* eslint-disable-next-line @typescript-eslint/no-explicit-any */ }
          { React.isValidElement<any>(action) ?
            action : (
              <a
                className={classNames(`Alert-${(type)}`, 'primary-action')}
                href={action.url}
                rel="noopener noreferrer"
                target={actionTarget}
              >
                {action.content}
              </a>
          )}
        </div>
      )
}
      {
        onDismiss && (
          <div className="Alert__close">
            <button
              aria-label={`close ${type}`}
              className="close"
              type="button"
              onClick={() => onDismiss(id)}
            >
              <FontAwesomeIcon icon={faTimes as IconDefinition} />
            </button>
          </div>
        )
      }
    </div>
  );
}

// Alert.propTypes = {
//   /**
//    Creates a CTA button on the Alert
//   */
//   action: PropTypes.oneOfType([PropTypes.object, PropTypes.node]),
//   /**
//     Specifies where to open the linked document
//   */
//   actionTarget: PropTypes.string,
//   /**
//    Determines whether the Alert will disappear automatically
//   */
//   autoDismiss: PropTypes.bool,
//   id: PropTypes.string.isRequired,
//   message: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
//   removeBorderLeft: PropTypes.bool,
//   title: PropTypes.string,
//   /**
//    One of the MessageTypes
//   */
//   type: PropTypes.oneOf(Object.values(MessageTypes)).isRequired,
//   onDismiss: PropTypes.func,
// };

// Alert.defaultProps = {
//   action: undefined,
//   actionTarget: undefined,
//   autoDismiss: false,
//   removeBorderLeft: false,
//   title: undefined,
//   onDismiss: undefined,
// };

export default React.memo(Alert);
