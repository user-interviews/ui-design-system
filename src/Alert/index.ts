import Alert, { MessageTypes } from './Alert';

export {
  Alert,
  MessageTypes,
};
