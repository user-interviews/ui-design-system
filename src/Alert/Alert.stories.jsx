import React, { Fragment } from 'react';
import { boolean, text, withKnobs } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';

import { Alert, MessageTypes } from 'src/Alert';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faGoogle } from '@fortawesome/free-brands-svg-icons';
import mdx from './Alert.mdx';

import 'bootstrap/dist/css/bootstrap.min.css';
import '../../scss/global.scss';

export default {
  title: 'Components/Alert',
  component: Alert,
  decorators: [withKnobs],
  parameters: {
    docs: {
      page: mdx,
    },
  },
};

export const Success = () => (
  <Alert
    id="1"
    message={text('Message', 'Success message')}
    removeBorderLeft={boolean('removeBorderLeft', false)}
    title={text('Title', 'Success title')}
    type={MessageTypes.SUCCESS}
  />
);

export const Info = () => (
  <Alert
    id="2"
    message={text('Message', 'Info message')}
    removeBorderLeft={boolean('removeBorderLeft', false)}
    title={text('Title', 'Info title')}
    type={MessageTypes.INFO}
  />
);

export const Feature = () => (
  <Alert
    id="3"
    message={text('Message', 'Some context around new feature if needed.')}
    removeBorderLeft={boolean('removeBorderLeft', false)}
    title={text('Title', 'New feature alert!')}
    type={MessageTypes.FEATURE}
  />
);

export const Error = () => (
  <Alert
    id="4"
    message={text('Message', 'Error message')}
    removeBorderLeft={boolean('removeBorderLeft', false)}
    title={text('Title', 'Error title')}
    type={MessageTypes.ERROR}
  />
);

export const Warning = () => (
  <Alert
    id="5"
    message={text('Message', 'Warning message')}
    removeBorderLeft={boolean('removeBorderLeft', false)}
    title={text('Title', 'Warning title')}
    type={MessageTypes.WARNING}
  />
);

const onDismiss = (id) => {
  action('alert dismissed')(id);
};

export const WithDismiss = () => (
  <Alert
    id="6"
    message={text('Message', 'Default message')}
    removeBorderLeft={boolean('removeBorderLeft', false)}
    title={text('Title', 'Default title')}
    type={MessageTypes.SUCCESS}
    onDismiss={onDismiss}
  />
);

const googleButtonStyle = {
  backgroundColor: '#4285F4',
  border: '1px solid #4285F4',
  borderRadius: '4px',
  fontSize: '0.875rem',
  fontWeight: 'bold',
  lineHeight: '1.25rem',
  margin: '16px 0',
  padding: '8px 16px',
  color: '#FFFFFF',
  whiteSpace: 'nowrap',
};

const GoogleCalendarButton = () => (
  <button style={googleButtonStyle} type="button">
    <FontAwesomeIcon icon={faGoogle} style={{ height: '1rem', marginRight: '0.5rem' }} />
    Connect Google Calendar
  </button>
);

export const WithCallToAction = () => (
  <>
    <p>Default CTA rendered by component</p>
    <Alert
      action={{ content: 'Primary action', url: 'https://www.userinterviews.com/' }}
      id="8"
      message="Success message"
      removeBorderLeft={boolean('removeBorderLeft', false)}
      title="Success title"
      type={MessageTypes.SUCCESS}
      onDismiss={onDismiss}
    />
    <Alert
      action={{ content: 'Primary action', url: 'https://www.userinterviews.com/' }}
      id="9"
      message="Info message"
      removeBorderLeft={boolean('removeBorderLeft', false)}
      title="Info title"
      type={MessageTypes.INFO}
      onDismiss={onDismiss}
    />
    <Alert
      action={{ content: 'Primary action', url: 'https://www.userinterviews.com/' }}
      id="11"
      message="Error message"
      removeBorderLeft={boolean('removeBorderLeft', false)}
      title="Error title"
      type={MessageTypes.ERROR}
      onDismiss={onDismiss}
    />
    <Alert
      action={{ content: 'Primary action', url: 'https://www.userinterviews.com/' }}
      id="12"
      message="Warning message"
      removeBorderLeft={boolean('removeBorderLeft', false)}
      title="Warning title"
      type={MessageTypes.WARNING}
      onDismiss={onDismiss}
    />
    <Alert
      action={{ content: 'Primary action', url: 'https://www.userinterviews.com/' }}
      id="13"
      message="Some context around new feature if needed"
      removeBorderLeft={boolean('removeBorderLeft', false)}
      title="New feature alert!"
      type={MessageTypes.FEATURE}
      onDismiss={onDismiss}
    />

    <p>Custom CTA via <code>message</code> prop</p>
    <Alert
      id="7"
      message={(
        <Fragment>
          <div>When you confirm a session we’ll automatically
            add an event and reminders to your Google Calendar.
          </div>
          <GoogleCalendarButton />
        </Fragment>
      )}
      removeBorderLeft={boolean('removeBorderLeft', false)}
      title={text('Title', 'Connect to Google Calendar to create reminders automatically')}
      type={MessageTypes.ANNOUNCEMENT}
      onDismiss={onDismiss}
    />

    <p>Custom CTA via <code>action</code> prop</p>
    <Alert
      action={<GoogleCalendarButton />}
      id="8"
      message={text('Message', `When you confirm a session we’ll automatically
      add an event and reminders to your Google Calendar.`)}
      removeBorderLeft={boolean('removeBorderLeft', false)}
      title={text('Title', 'Connect to Google Calendar to create reminders automatically')}
      type={MessageTypes.ANNOUNCEMENT}
      onDismiss={onDismiss}
    />
  </>
);
