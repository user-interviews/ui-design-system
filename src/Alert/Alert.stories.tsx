import React, { Fragment } from 'react';
import { action } from '@storybook/addon-actions';

import { Alert, MessageTypes } from 'src/Alert';

import mdx from './Alert.mdx';

export default {
  title: 'Components/Alert',
  component: Alert,
  parameters: {
    docs: {
      page: mdx,
    },
  },
};

export function Success() {
  return (
    <Alert
      id="1"
      message="Success message"
      removeBorderLeft={false}
      title="Success title"
      type={MessageTypes.SUCCESS}
    />
  );
}

export function Info() {
  return (
    <Alert
      id="2"
      message="Info message"
      removeBorderLeft={false}
      title="Info title"
      type={MessageTypes.INFO}
    />
  );
}

export function Feature() {
  return (
    <Alert
      id="3"
      message="Some context around new feature if needed."
      removeBorderLeft={false}
      title="New feature alert!"
      type={MessageTypes.FEATURE}
    />
  );
}

export function Error() {
  return (
    <Alert
      id="4"
      message="Error message"
      removeBorderLeft={false}
      title="Error title"
      type={MessageTypes.ERROR}
    />
  );
}

export function Warning() {
  return (
    <Alert
      id="5"
      message="Warning message"
      removeBorderLeft={false}
      title="Warning title"
      type={MessageTypes.WARNING}
    />
  );
}

const onDismiss = (id) => {
  action('alert dismissed')(id);
};

export function WithDismiss() {
  return (
    <Alert
      id="6"
      message="Default message"
      removeBorderLeft={false}
      title="Default title"
      type={MessageTypes.SUCCESS}
      onDismiss={onDismiss}
    />
  );
}

export function WithCallToAction() {
  return (
    <>
      <p>Default CTA rendered by component</p>
      <Alert
        action={{ content: 'Primary action', url: 'https://www.userinterviews.com/' }}
        id="8"
        message="Success message"
        removeBorderLeft={false}
        title="Success title"
        type={MessageTypes.SUCCESS}
        onDismiss={onDismiss}
      />
      <Alert
        action={{ content: 'Primary action', url: 'https://www.userinterviews.com/' }}
        id="9"
        message="Info message"
        removeBorderLeft={false}
        title="Info title"
        type={MessageTypes.INFO}
        onDismiss={onDismiss}
      />
      <Alert
        action={{ content: 'Primary action', url: 'https://www.userinterviews.com/' }}
        id="11"
        message="Error message"
        removeBorderLeft={false}
        title="Error title"
        type={MessageTypes.ERROR}
        onDismiss={onDismiss}
      />
      <Alert
        action={{ content: 'Primary action', url: 'https://www.userinterviews.com/' }}
        id="12"
        message="Warning message"
        removeBorderLeft={false}
        title="Warning title"
        type={MessageTypes.WARNING}
        onDismiss={onDismiss}
      />
      <Alert
        action={{ content: 'Primary action', url: 'https://www.userinterviews.com/' }}
        id="13"
        message="Some context around new feature if needed"
        removeBorderLeft={false}
        title="New feature alert!"
        type={MessageTypes.FEATURE}
        onDismiss={onDismiss}
      />
    </>
  );
}
