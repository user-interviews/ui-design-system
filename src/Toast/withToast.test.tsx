import React from 'react';
import { act } from '@testing-library/react-hooks';
import { create } from 'react-test-renderer';

import { MessageTypes } from '../Alert';
import { withToast } from '.';

jest.mock('react-transition-group', () => (
  {
    CSSTransition: ({ children }) => children,
    TransitionGroup: ({ children }) => children,
  }
));

const WrappedComponent = () => <div />;

describe('test withToast', () => {
  test('it can create a new Toast message', async () => {
    const newMessage = 'This is just a test...';
    const ComponentWithToast = withToast(WrappedComponent);
    const toast = create(<ComponentWithToast />);
    const component = toast.root.findByType(WrappedComponent);

    act(() => {
      component.props.setToastMessage({
        type: MessageTypes.SUCCESS,
        message: newMessage,
      });
    });

    expect(toast).toMatchSnapshot();
  });
});
