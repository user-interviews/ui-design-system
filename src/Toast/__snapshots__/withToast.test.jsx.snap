// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test withToast it can create a new Toast message 1`] = `
Array [
  <div
    className="Alert Alert-success"
  >
    <div
      className="Alert__icon"
    >
      <span
        className="fa-layers fa-fw"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-circle fa-w-16 "
          data-icon="circle"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={
            Object {
              "transformOrigin": "0.5em 0.5em",
            }
          }
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            style={Object {}}
            transform="translate(256 256)"
          >
            <g
              style={Object {}}
              transform="translate(0, 0)  scale(1.5, 1.5)  rotate(0 0 0)"
            >
              <path
                d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8z"
                fill="currentColor"
                style={Object {}}
                transform="translate(-256 -256)"
              />
            </g>
          </g>
        </svg>
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-check fa-w-16 "
          data-icon="check"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={
            Object {
              "transformOrigin": "0.5em 0.5em",
            }
          }
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            style={Object {}}
            transform="translate(256 256)"
          >
            <g
              style={Object {}}
              transform="translate(0, 0)  scale(0.75, 0.75)  rotate(0 0 0)"
            >
              <path
                d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
                fill="currentColor"
                style={Object {}}
                transform="translate(-256 -256)"
              />
            </g>
          </g>
        </svg>
      </span>
    </div>
    <div
      className="Alert__content"
    >
      This is just a test...
    </div>
    <div
      className="Alert__close"
    >
      <button
        aria-label="close"
        className="close"
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-times fa-w-11 "
          data-icon="times"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </div>
  </div>,
  <div />,
]
`;
